### DISCUSSION TITLE
Modulators send to midi & audio tracks

### DISCUSSION THREAD

**Discussion by Ponchomouse** [SynthstromAudible/DelugeFirmware#2711] [Ideas]
Loving the latest build. Some truly unique features. My deluge is turning into the main hub for everything but one feature I feel is missing is sending things like the lfo to a midi cc, maybe even a envelope to the cc. Also be good to be able to have them affect the effects on an audio track as well. I end up importing a audio track to a kit so I can do that type of thing. 

I have an nts1 with all the midi cc mapped to the gold knobs but can only manually record automation that way

Not sure if this is a tricky ask though given the lfo is part of the synth side of things.

Many thanks

**Comment by duncanrmi** [comment]
I would think it would be do-able. 

the LFOs are just 0s & 1s whizzing around inside the CPU after all.& the mod matrix can cope ok with _incoming_ CCs.

much lowlier devices, no disrespect intended, such as the NDLR, can route their internal LFOs onto CC as well as the internal workings.

**Comment by Qui-Gon-JooL** [comment]
When I bought the deluge I was convinced that the modulators, like the lfo and the envelopes worked in midi out, to control external CC, like the Elektron does, I had not inquired about this point before buying it. bought, it seemed obvious to me. 
 I was very surprised and disappointed to find that no... so much so that I almost sold it directly which would have been a big mistake given how much I love it.

I don't know if this is physically possible? Maybe the modulators are coded specifically for the synth and impossible to send midi... But actually I would love that, the fact of being able to have an almost unlimited number of lfo and envelope...

**Reply by RJ-Eckie** [reply]
> I don't know if this is physically possible? Maybe the modulators are coded specifically for the synth and impossible to send midi... But actually I would love that, the fact of being able to have an almost unlimited number of lfo and envelope...

Envelopes are generally too fast to send over midi at all - Elektron doesn't do this either. From the Mk1 devices (digitakt, digitone, syntakt) I also know that the midi tracks only have one lfo, while all regular tracks have two. So yes it's possible, but especially over regular (non-USB) midi there are some limitations. I still think this would be a very interesting feature to implement, but it would have to be in a smart way to deal with those limitations. Midi over USB could possibly be a lot faster, but this would also have to be implemented in a smart way - currently there are no differences between midi tracks being used over hardware midi or usb midi.

As Sean writes below, apparently the Deluge code isn't very optimized in this regard at this point. (I haven't looked into it specifically yet.)

**Reply by Qui-Gon-JooL** [reply]
When I was talking about Elektron envelopes I was talking about LFOs which can be defined on a single cycle and which by definition transforms them into an envelope.

It is true that the MK1 only has one LFO per midi track, but I have already reinjected the 7 midi tracks of the Digitakt internally in/out to only track 1 to modulate all the parameters of a single sampler track and It worked without a bug, so I think it's a deliberate limitation and not a material one.

I think that outputting the MIDI lfos via USB MIDI would not be of much interest since it is mainly a question of sending these data to another machine or to the internal synth (MIDI loop back) in a case of dawless. 
If we send the midi via usb there is a good chance that we are talking about a computer and the LFO solution is not lacking..

**Reply by RJ-Eckie** [reply]
The Deluge can midi-host up to 4 hardware devices over its USB port so that does make a lot of sense to me

**Reply by Qui-Gon-JooL** [reply]
Ah in this case ok, but what kind of material are you referring to? An example? Because I have various synth and groove boxes and none of them can receive midi via the Deluge's USB.

**Reply by RJ-Eckie** [reply]
I don't have a list of these memorized, but any device that if you connect it to your computer and you can send it midi from your daw without installing any drivers, should also work on the Deluge. Two examples I own are the Roland Aira S-1 and the Korg NTS-3. My stage piano (Nord Grand) should also work, although I don't use it together with my Deluge. Basically any device with class compliant midi should work - but again I haven't tested a wide range so I'm just telling you what I know from the documentation

**Comment by seangoodvibes** [comment]
> When I bought the deluge I was convinced that the modulators, like the lfo and the envelopes worked in midi out, to control external CC, like the Elektron does, I had not inquired about this point before buying it. bought, it seemed obvious to me. 
> 
>  I was very surprised and disappointed to find that no... so much so that I almost sold it directly which would have been a big mistake given how much I love it.
> 
> 
> 
> I don't know if this is physically possible? Maybe the modulators are coded specifically for the synth and impossible to send midi... But actually I would love that, the fact of being able to have an almost unlimited number of lfo and envelope...

It's not impossible, but imagine if you were sending that much midi cc data out the deluge. You would crash the midi connection.

I really don't know how devices like Elektron, Hapax handle sending out midi lfo data, but the one thing the Deluge needs to get better at is throttling the speed of midi data going out because the midi protocol can only handle so much midi bandwidth.

As an example, just try going into a midi clip and write some automation at 128th resolution. Create some shapes with every step at that resolution. Then send the midi out.

Watch as your deluge will probably lag like hell or potentially even crash. Not sure what would happen to the receiving device either.

We could add lfo's and stuff that sends at a higher resolution. But it won't be a true lfo in the sense of smooth precise curves. It would be a stepped lfo.

**Reply by Qui-Gon-JooL** [reply]
It's true, I've already experienced midi slowdowns in this kind of case, in fact a step-by-step lfo would be great, even if only the "random mod" could get out of midi it would already be great in fact.

**Comment by darwiniandude** [comment]
I have a Squarp Pyramid. Sure it's a dedicated sequencer, but with non velocity sensitive pads it has some overlap with Deluge. It has MIDI LFOs. I would love the option of LFO's for midi tracks. Lots of external gear from all eras have midi CC#'s to control parameters, and being able to easily modulate these ads a lot. Should go to CV too. Not trying to make the Deluge over complicated, it doesn't need all this: but the way Pyramid handles it is 5 midi effect insert slots on a track. You can add a harmonizer or and LFO for example. You can chain LFO's to combine them into a different midi CC LFO waveform, eg combining a triangle and sine. Anyway - some sort of midi LFO output would be cool on Deluge :)
